basetest.pm

```diff
@@ -164,7 +164,18 @@ sub record_screenmatch {
     if ($foundneedle->has_property('workaround')) {
         $result->{dent} = 1;
         $self->{dents}++;
-        bmwqemu::diag("needle '$serialized_match->{name}' is a workaround");
+
+        # get the workaround reason and write a test result file.
+        my $reason = $foundneedle->get_property_value('workaround');
+
+        $result->{title} = 'Soft Failed';
```

Comment:
This is copied and pasted from `record_soft_failure`. Why don't you simply make use of that function? Of course you need to remove `$result->{dent} = 1;` and `$self->{dents}++;` at the top of the `if`-block because `$self->record_soft_failure($reason)` would already take care of that.

Reply:
record_soft_failure is a function in testapi.pm, could I call this in basetest.pm? There is a function called "record_soft_failure_result", maybe you mean it? but it doesn't do $self->{dents}++ and $result->{dent}. So there is a little different.

Reply:
I mean `record_soft_failure_result` and my previous comment already contains a note about `$self->{dents}++ and $result->{dent}`.

Reply:
I still have some questions about those code:
a. In you previous comment, you said I need to remove `result->{dent} = 1` and `$self->{dents}++;`, I do not understand clearly what 'dent' and 'dents' used for.   I cannot find out where `record_soft_failure_result` has handled them.
b. function `record_soft_failure_result` calls `record_testresult`, it does `++self->{test_count}`, as a result, the next result file is 1 more than expected (In my opinion). For example, In my testcase, when I call `record_soft_failure_result`, the result file name is 'basetest-3.txt'; when I do not call it, the result file name is 'basetest-2.txt'.
c.  Function `record_soft_failure_result` calls `record_testresult`, and `record_testresult` push a result hash to `$self->{details}`. In the end of function `record_screenmatch` also push a result hash to `$self->{details}`, it repeats.
I really did test for these. This is why I did not simply call `record_soft_failure_result` in my code. Maybe I made some mistakes in understanding the 'record_screenmatch' function. If I did, please let me know.

Reply:
> I cannot find out where record_soft_failure_result has handled them.

Those are actually handled in `testapi::record_softfailure` and `testapi::force_soft_failure` and *not* inside `basetest`. But maybe we want to change this inconsistency and increase the dents directly in `basetest::record_soft_failure_result`? That would make sense to me. Otherwise you actually need to keep the dents in place.

---

> I do not understand clearly what 'dent' and 'dents' used for.

To count the number of issues which have been occurred during the test run. Dents is actually (in my opinion) a poorly chosen word in that context. I think someone was a bit too eager using sophisticated vocabulary.

---

So I admit that using `basetest::record_soft_failure_result` is not that easily possible. Beside from that: Your code marks the screen match itself as the softfailure which make indeed sense. To store the reason you then use `record_serialresult`. That on the other hand is actually not so nice because it looks like `wait_serial` was used in the test code.

What you actually want here is to combine different result types (screenshot and text) but to my knowledge that's not possible (without adjusting also the frontend and making things even more complicated). Maybe the simplest solution is to mark the screenmatch result as softfail by setting `$result->{result} = 'softfail';` as your code already does and then record the reason as a separate text result just as if `record_softfailure` was called. That could be implemented by making use of `record_soft_failure_result`.

If you decide to move increment the dents to basetest which I would advocate you can get rid of increasing the dents for softfailure screenmatch.

---

> function record_soft_failure_result calls record_testresult, it does ++self->{test_count}, as a result, the next result file is 1 more than expected (In my opinion).

That *might* be a problem, but have you tested how the fronted handles actully that? That the frontend can display the results correctly is what matters.  
By the way, it would make sense to share a test run with such a softfailure (e.g. just a screenshot so we see how it looks like).